File: android/app/src/main/java/ai/mapier/swipe/analytics/AnalyticsEvent.kt
package ai.mapier.swipe.analytics

enum class AnalyticsProvider(val value: String) {
  SPOTIFY("spotify"),
  APPLE_MUSIC("apple_music"),
}

enum class AnalyticsErrorCategory(val value: String) {
  AUTHORIZATION("authorization"),
  ELIGIBILITY("eligibility"),
  LIBRARY_LOAD("library_load"),
  COMMIT("commit"),
  PLAYBACK("playback"),
  CONFIGURATION("configuration"),
  UNKNOWN("unknown"),
}

sealed interface AnalyticsEvent {
  val name: String
  val properties: Map<String, String>

  data object AppOpened : AnalyticsEvent {
    override val name = "app_opened"
    override val properties = emptyMap<String, String>()
  }

  data object ProviderPickerViewed : AnalyticsEvent {
    override val name = "provider_picker_viewed"
    override val properties = emptyMap<String, String>()
  }

  data class ProviderConnectionStarted(val provider: AnalyticsProvider) : AnalyticsEvent {
    override val name = "provider_connection_started"
    override val properties = provider.properties
  }

  data class ProviderConnectionSucceeded(val provider: AnalyticsProvider) : AnalyticsEvent {
    override val name = "provider_connection_succeeded"
    override val properties = provider.properties
  }

  data class ProviderConnectionFailed(
    val provider: AnalyticsProvider,
    val category: AnalyticsErrorCategory,
  ) : AnalyticsEvent {
    override val name = "provider_connection_failed"
    override val properties = provider.properties + ("error_category" to category.value)
  }

  data class CleanupDeckLoaded(val provider: AnalyticsProvider) : AnalyticsEvent {
    override val name = "cleanup_deck_loaded"
    override val properties = provider.properties
  }

  data class FirstDecisionCompleted(val provider: AnalyticsProvider) : AnalyticsEvent {
    override val name = "first_decision_completed"
    override val properties = provider.properties
  }

  data class ReviewOpened(val provider: AnalyticsProvider) : AnalyticsEvent {
    override val name = "review_opened"
    override val properties = provider.properties
  }

  data class CleanupSessionCompleted(val provider: AnalyticsProvider) : AnalyticsEvent {
    override val name = "cleanup_session_completed"
    override val properties = provider.properties
  }

  data class CleanupSessionAbandoned(val provider: AnalyticsProvider) : AnalyticsEvent {
    override val name = "cleanup_session_abandoned"
    override val properties = provider.properties
  }

  data object AccountDeleted : AnalyticsEvent {
    override val name = "account_deleted"
    override val properties = emptyMap<String, String>()
  }
}

private val AnalyticsProvider.properties: Map<String, String>
  get() = mapOf("provider" to value)

fun interface AnalyticsCapturing {
  fun capture(event: AnalyticsEvent)
}

object NoOpAnalytics : AnalyticsCapturing {
  override fun capture(event: AnalyticsEvent) = Unit
}

class AnalyticsClient(
  private val captureEvent: (String, Map<String, String>) -> Unit,
) : AnalyticsCapturing {
  override fun capture(event: AnalyticsEvent) {
    captureEvent(event.name, event.properties)
  }
}

data class AnalyticsSettings(
  val projectToken: String,
  val host: String,
) {
  val isEnabled: Boolean
    get() = projectToken.isNotBlank() && runCatching {
      val uri = java.net.URI(host)
      uri.scheme == "https" && !uri.host.isNullOrBlank()
    }.getOrDefault(false)
}

File: ios/Wavepoint/Analytics/AnalyticsEvent.swift
import Foundation

enum AnalyticsProvider: String, Sendable {
  case spotify
  case appleMusic = "apple_music"

  init(_ provider: MusicProvider) {
    self = provider == .spotify ? .spotify : .appleMusic
  }
}

enum AnalyticsErrorCategory: String, Sendable {
  case authorization
  case eligibility
  case libraryLoad = "library_load"
  case commit
  case playback
  case configuration
  case unknown
}

enum AnalyticsEvent: Equatable, Sendable {
  case appOpened
  case providerPickerViewed
  case providerConnectionStarted(AnalyticsProvider)
  case providerConnectionSucceeded(AnalyticsProvider)
  case providerConnectionFailed(AnalyticsProvider, AnalyticsErrorCategory)
  case cleanupDeckLoaded(AnalyticsProvider)
  case firstDecisionCompleted(AnalyticsProvider)
  case reviewOpened(AnalyticsProvider)
  case cleanupSessionCompleted(AnalyticsProvider)
  case cleanupSessionAbandoned(AnalyticsProvider)
  case accountDeleted

  var name: String {
    switch self {
    case .appOpened: "app_opened"
    case .providerPickerViewed: "provider_picker_viewed"
    case .providerConnectionStarted: "provider_connection_started"
    case .providerConnectionSucceeded: "provider_connection_succeeded"
    case .providerConnectionFailed: "provider_connection_failed"
    case .cleanupDeckLoaded: "cleanup_deck_loaded"
    case .firstDecisionCompleted: "first_decision_completed"
    case .reviewOpened: "review_opened"
    case .cleanupSessionCompleted: "cleanup_session_completed"
    case .cleanupSessionAbandoned: "cleanup_session_abandoned"
    case .accountDeleted: "account_deleted"
    }
  }

  var properties: [String: String] {
    switch self {
    case .providerConnectionFailed(let provider, let category):
      ["provider": provider.rawValue, "error_category": category.rawValue]
    case .providerConnectionStarted(let provider),
      .providerConnectionSucceeded(let provider),
      .cleanupDeckLoaded(let provider),
      .firstDecisionCompleted(let provider),
      .reviewOpened(let provider),
      .cleanupSessionCompleted(let provider),
      .cleanupSessionAbandoned(let provider):
      ["provider": provider.rawValue]
    case .appOpened, .providerPickerViewed, .accountDeleted:
      [:]
    }
  }
}

protocol AnalyticsCapturing: Sendable {
  func capture(_ event: AnalyticsEvent)
}

struct NoOpAnalytics: AnalyticsCapturing {
  func capture(_: AnalyticsEvent) {}
}

final class AnalyticsClient: AnalyticsCapturing, @unchecked Sendable {
  private let captureEvent: (String, [String: String]) -> Void

  init(captureEvent: @escaping (String, [String: String]) -> Void) {
    self.captureEvent = captureEvent
  }

  func capture(_ event: AnalyticsEvent) {
    captureEvent(event.name, event.properties)
  }
}

struct AnalyticsSettings: Equatable, Sendable {
  let projectToken: String
  let host: String

  var isEnabled: Bool {
    guard !projectToken.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
      let url = URL(string: host),
      url.scheme == "https",
      url.host != nil
    else {
      return false
    }
    return true
  }
}
